lightning/src/ln/outbound_payment.rs

```diff
@@ -1006,36 +1024,58 @@ impl OutboundPayments {
 		keysend_preimage: Option<PaymentPreimage>, route: &Route, retry_strategy: Option<Retry>,
 		payment_params: Option<PaymentParameters>, entropy_source: &ES, best_block_height: u32
 	) -> Result<Vec<[u8; 32]>, PaymentSendFailure> where ES::Target: EntropySource {
+		let mut pending_outbounds = self.pending_outbound_payments.lock().unwrap();
+		let entry = pending_outbounds.entry(payment_id);
+		if let hash_map::Entry::Occupied(entry) = &entry {
+			if let PendingOutboundPayment::AwaitingInvoice { .. } = entry.get() { } else {
+				return Err(PaymentSendFailure::DuplicatePayment)
```

Comment:
Could you describe the scenario in more detail? Note that there is a `send_payment_for_bolt12_invoice` method added in the next PR.

Reply:
Ah, I see... Unfortunately, while trying to reuse `send_payment_internal`, `add_new_pending_payment` would be called in both cases, so this will need some refactoring unless we are ok duplicating code.

It would be nice to perform the check upfront such that the code can be reused still without needing to do something like pipe a `is_bol12` parameter through the helpers. But even so, it seems we don't hold the `pending_outbound_payments` lock throughout these calls. So there's probably edge cases where the check is valid initially but may not be valid later if the locks isn't held throughout.

Would it make sense to hold on to the locks and pass a hash map entry (either occupied or vacant) to the helper functions? Or is there a reason why we drop and re-acquire the lock throughout these helpers? (cc: @valentinewallace )

Reply:
Ah, that makes sense. Ok, reverted but will be in the next PR as per https://github.com/lightningdevkit/rust-lightning/pull/2468#discussion_r1307780611.